### PR TITLE
nrunner: exports to the tests a temporary output dir (and a small refactoring)

### PR DESCRIPTION
**Pull request by beraldoleal**
This pull request will introduce the ability to tests to see through the environment variable AVOCADO_TEST_OUTPUT_DIR where to write output results. This is part of #3695 and is the foundation to collect those results. Also, a small organization is proposed here, removing spawners from nrunner.py.

### CODE REVIEW

File: avocado/core/nrunner.py [hunk 1]
```diff
@@ -279,6 +279,26 @@ class BaseRunner:
...
+        :param: status_type: The type of event ('started', 'running',
+                             'finished')
+        :param: addional_info: Any addional information that you
+        would like to
+                               add to the dict. This must be a dict.
+
+        :rtype: dict
+        """
+        status = {'status': status_type,
+                  'time': time.time()}
```
Comment (clebergnu) [inline avocado/core/nrunner.py:296]: The `tap` runner currently reports a `timestamp` that is identical to this. We need to either use one or the other.

File: avocado/core/nrunner.py [hunk 2]
```diff
@@ -424,23 +434,22 @@ def _run_unittest(uri, queue):
         output = {'status': 'finished',
                   'result': result,
                   'output': stream.read(),
-                  'time_end': time_end}
+                  'time': time_end}
```
Comment (clebergnu) [inline avocado/core/nrunner.py:438]: This will be overwritten by https://github.com/avocado-framework/avocado/pull/3723/commits/12488c9a5508e9a6e50a4903c6cc157e7cab2c3f#diff-5a2f92aedf41bef8d89fc436e1b3c84cR296, won't it?

We have to decide if we'll let the subclasses have the final saying in the time (like in this example), and I think they should, as this would be more accurate.

File: avocado/core/nrunner.py [hunk 3]
```diff
@@ -583,6 +586,7 @@ def __init__(self, identifier, runnable, status_uris=None, known_runners=None):
             known_runners = {}
         self.known_runners = known_runners
         self.spawn_handle = None
+        self._output_dir = "/tmp/avocado-tasks/{}/output".format(uuid4())
```
Comment (clebergnu) [inline avocado/core/nrunner.py:595]: What's the advantage you see in "cooking" our own directory location, instead of using Python's standard library support? https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

File: selftests/unit/test_nrunner.py [hunk 4]
```diff
@@ -191,7 +191,7 @@ def test_runner_noop(self):
         results = [status for status in runner.run()]
         last_result = results[-1]
         self.assertEqual(last_result['status'], 'finished')
-        self.assertIn('time_end', last_result)
+        self.assertIn('time', last_result)
```
Comment (clebergnu) [inline selftests/unit/test_nrunner.py:194]: These (4 in total) don't belong in this commit.

File: avocado/core/nrunner.py (hunk 1 above)
Comment (beraldoleal) [reply avocado/core/nrunner.py:296]: @clebergnu that is right. my fault, I was not paying attention to the other runners. I will change it during v2.

File: avocado/core/spawners/podman.py [hunk 5]
```diff
@@ -71,3 +71,9 @@ def spawn_task(self, task):
                                                          stdout=asyncio.subprocess.PIPE,
                                                          stderr=asyncio.subprocess.PIPE)
         yield from proc.wait()
+
+    def create_output_dir_for_task(self, task):
+        """In this case is a remote folder."""
+        local_dir = '/tmp/avocado/tasks/{}/output'.format(task.identifier)
```
Comment (clebergnu) [inline avocado/core/spawners/podman.py:77]: I am really confused here. You're creating a directory locally, and then asking the task to create it... is this intended to be the destination mount point for a `-v` usage?

File: avocado/core/spawners/podman.py [hunk 6]
```diff
@@ -71,3 +71,9 @@ def spawn_task(self, task):
...
+        os.mkdirs(local_dir, exist_ok=True)
+        task.set_output_dir(local_dir)
```
Comment (clebergnu) [inline avocado/core/spawners/podman.py:79]: I don't see a `set_output_dir()` anywhere but here.

File: avocado/core/nrunner.py (hunk 2 above)
Comment (beraldoleal) [reply avocado/core/nrunner.py:438]: @clebergnu, to avoid having some runners without the information or with a different key ('time' vs 'timestamp' vs 'time_start' vs 'time_end')  I thought that is worthed to create a convention here: *Every message will have the 'time' key.*

I understand that if we let this in the subclass we will have a "more accurate" time, but I did a few runs comparing both, and this is the difference that we are talking about:

```
direct: 1587139446.4457245
indirect: 1587139446.4457424
```

Unless that I'm missing something, I think that putting this inside a common method will simplify the code and ensure that all messages will have the time (in case some runner forgot about it). Also with a common method is easy to add default values in the future for specific cases if needed, only one place change.

But I confess that I'm not sure if this small difference in time will impact here. What do you think?

File: avocado/core/nrunner.py (hunk 3 above)
Comment (beraldoleal) [reply avocado/core/nrunner.py:595]: IIUC, TemporaryDirectory will try to remove the directory when the context is finished. And I understood that we would like to keep those dirs, right?

We could use `mkdtemp` that is not destructive, but this is creating when called and since that, we are sending the possible directory name during the "start" maybe this is not suitable otherwise all Task instance would have an empty dir left.

I tried to create only when `run()` is called, but having the name "a priori". But for sure this logic can be changed this logic if you think that is valid. What do you think?

File: selftests/unit/test_nrunner.py (hunk 4 above)
Comment (beraldoleal) [reply selftests/unit/test_nrunner.py:194]: You are right, fixed in v2.

File: avocado/core/spawners/podman.py (hunk 5 above)
Comment (beraldoleal) [reply avocado/core/spawners/podman.py:77]: Sorry, my bad, this is a leftover of a rebase. I will remove this method. Is not being used anymore.

File: avocado/core/spawners/podman.py (hunk 6 above)
Comment (beraldoleal) [reply avocado/core/spawners/podman.py:79]: Yes, leftover, sorry again.

File: avocado/core/nrunner.py (hunk 2 above)
Comment (clebergnu) [reply avocado/core/nrunner.py:438]: > @clebergnu, to avoid having some runners without the information or with a different key ('time' vs 'timestamp' vs 'time_start' vs 'time_end') I thought that is worthed to create a convention here: _Every message will have the 'time' key._
> 

This is fine, and it's a good idea. Still, it doesn't conflict with some runner subclasses wanting to provide the time themselves, right? IMO, It's just a matter of the base class method not overwriting something if the subclass set it.

> I understand that if we let this in the subclass we will have a "more accurate" time, but I did a few runs comparing both, and this is the difference that we are talking about:
> 
> ```
> direct: 1587139446.4457245
> indirect: 1587139446.4457424
> ```
> 

This is a very small difference, on such a small test and on an underloaded system. It may be different, and relevant, on different test environment.

> Unless that I'm missing something, I think that putting this inside a common method will simplify the code and ensure that all messages will have the time (in case some runner forgot about it). Also with a common method is easy to add default values in the future for specific cases if needed, only one place change.
> 
> But I confess that I'm not sure if this small difference in time will impact here. What do you think?

I just think a subclass should have the final say about its status messages and the `time` is just one example.  I do agree that is not very important right now because we don't have tests running on loaded system reporting, say, +/-30% different execution times, so it's really about the hypothetical issue.

Other than that, do you think that a runner passing `time` to `prepare_status` and finding that it got changed is correct, implementation wise?

**Comment by beraldoleal** [comment]
@clebergnu thanks for your review. I did a mess-up with a rebase, sorry. Will be fixed in v2.

I replied to all your comments here, please let me know what you think on them, before I submmit a v2.

File: avocado/core/nrunner.py (hunk 2 above)
Comment (beraldoleal) [reply avocado/core/nrunner.py:438]: > This is fine, and it's a good idea. Still, it doesn't conflict with some runner subclasses wanting to provide the time themselves, right? IMO, It's just a matter of the base class method not overwriting something if the subclass set it.

Yes, I agree. And I can add an IF to base class to check if there is no 'time'. 

One problem that I can see with this approach is the lack of a "convention". "base class" can try to see if `time` is set but the runner can use a different name like `my_custom_timer`. So everything is about a convention and in order to avoid that I was trying to centralize the things here.

> Other than that, do you think that a runner passing `time` to `prepare_status` and finding that it got changed is correct, implementation wise?

No, it isn't correct. And that was, not the idea. The idea is to not pass `time` to `prepare_status`, but `prepare_status` create the time for the runner. The runner will say: give-me a status_message. Like "generate an event".

Comment (clebergnu) [reply avocado/core/nrunner.py:438]: Right, so I think you can (instead of an `if`) just start with the status the subclass is giving and `dict.update()` it with the information that is set by convention/default on `prepare_status()`.

File: avocado/core/nrunner.py (hunk 3 above)
Comment (clebergnu) [reply avocado/core/nrunner.py:595]: The fact that you're hoping to not collide based on the UUID before creating the directory is going to be safe on, say, more than 99.99999999999% of times, but it's *not* proven to be free of a race condition.

TBH, I'd rather deal with this issue of correctness, even if it requires extra code.

File: avocado/core/nrunner.py (hunk 2 above)
Comment (beraldoleal) [reply avocado/core/nrunner.py:438]: @clebergnu IIUC with your suggestion will bring back the problem that you point me out: the base class will override the time provided by the runner.

```python
default = {'time': 2} # base class
additional_info = {'time': 1} # subclass (runner)
addititonal_info.update(default)
print(additional_info['time'])
2
```

Inverting this logic (start with default and update with that is provided by the runner) is the current implementation.

File: avocado/core/nrunner.py (hunk 3 above)
Comment (beraldoleal) [reply avocado/core/nrunner.py:595]: @clebergnu Yes, I agree. I tought that the 0.00000001% collision problem would be safe, but fair enought, let's change.

File: avocado/core/nrunner.py (hunk 2 above)
Comment (clebergnu) [reply avocado/core/nrunner.py:438]: > @clebergnu IIUC with your suggestion will bring back the problem that you point me out: the base class will override the time provided by the runner.
> 
> ```python
> default = {'time': 2} # base class
> additional_info = {'time': 1} # subclass (runner)
> addititonal_info.update(default)
> print(additional_info['time'])
> 2
> ```
> 
> Inverting this logic (start with default and update with that is provided by the runner) is the current implementation.

I was just testing your attention! :smile: 

Now, seriously, I meant that but clearly got it wrong when trying to explain it.

**Comment by beraldoleal** [comment]
Closing in favor of a v2.
